File: cpp/include/mass_loss_test.cpp
#include <iostream>
#include <cassert>
#include "mass_loss.hpp"

using namespace hot_jupiter;

int main() {
    std::cout << "[Unit Test] Mass Loss (RLOF & Photoevaporation)..." << std::endl;
    RocheLobeMassLoss rlof;
    double r_roche = RocheLobeMassLoss::roche_lobe_radius(0.02 * AU, 1.0 * M_JUP, 1.0 * M_SUN);
    assert(r_roche > 0.0 && r_roche < 0.02 * AU);

    double fill_under = rlof.roche_lobe_filling_factor(0.8 * r_roche, 0.02 * AU, 1.0 * M_JUP, 1.0 * M_SUN);
    double fill_over = rlof.roche_lobe_filling_factor(1.1 * r_roche, 0.02 * AU, 1.0 * M_JUP, 1.0 * M_SUN);
    assert(fill_under < 1.0);
    assert(fill_over > 1.0);

    double dM_dt_xuv = rlof.compute_photoevaporative_mdot(10.0, 1.4 * R_JUP, 1.0 * M_JUP);
    assert(dM_dt_xuv < 0.0);

    auto [dM_dt, da_dt] = rlof.evaluate_mass_loss_rate(1.05 * r_roche, 0.018 * AU, 1.0 * M_JUP, 1.0 * M_SUN, 5.0);
    assert(dM_dt < 0.0);
    assert(da_dt != 0.0);

    std::cout << "  -> PASSED." << std::endl;
    return 0;
}

File: hot_jupiter/mass_loss_test.py
"""
Unit tests for Roche Lobe Overflow (RLOF) mass loss module (hot_jupiter/mass_loss.py).
"""

from hot_jupiter.constants import AU, M_JUP, M_SUN
from hot_jupiter.mass_loss import RocheLobeMassLoss


def test_roche_lobe_radius():
    r_roche = RocheLobeMassLoss.roche_lobe_radius(a=0.02 * AU,
                                                  M_p=1.0 * M_JUP,
                                                  M_star=1.0 * M_SUN)
    assert r_roche > 0.0
    assert r_roche < 0.02 * AU


def test_roche_lobe_mass_loss_rate():
    rlof = RocheLobeMassLoss()
    a = 0.02 * AU
    M_p = 1.0 * M_JUP
    M_star = 1.0 * M_SUN

    r_roche = rlof.roche_lobe_radius(a, M_p, M_star)

    # Planet comfortably inside Roche Lobe -> 0 mass loss
    dM_dt, da_dt = rlof.evaluate_mass_loss_rate(R_p=0.5 * r_roche,
                                                a=a,
                                                M_p=M_p,
                                                M_star=M_star)
    assert dM_dt == 0.0
    assert da_dt == 0.0

    # Planet overflowing Roche Lobe (R_p > R_Roche) -> Negative mass loss
    dM_dt_overflow, da_dt_overflow = rlof.evaluate_mass_loss_rate(R_p=1.1 *
                                                                  r_roche,
                                                                  a=a,
                                                                  M_p=M_p,
                                                                  M_star=M_star)
    assert dM_dt_overflow < 0.0
    assert da_dt_overflow > 0.0
